src/context/chartLayoutContext.tsx

```diff
@@ -1,29 +1,98 @@
+import type { Props as YAxisProps } from '../cartesian/YAxis';
+import { calculateViewBox } from '../util/calculateViewBox';
 
-type ChartLayoutContextInitializing = {
-  xAxisMap: null;
-  yAxisMap: null;
-  viewBox: null;
-  clipPathId: null;
-};
+export const XAxisContext = createContext<XAxisMap | undefined>(undefined);
+export const YAxisContext = createContext<YAxisMap | undefined>(undefined);
+export const ViewBoxContext = createContext<CartesianViewBox | undefined>(undefined);
+export const ClipPathIdContext = createContext<string | undefined>(undefined);
 
-type ChartLayoutContextTypeLoaded = {
-  xAxisMap: XAxisMap;
-  yAxisMap: YAxisMap;
-  viewBox: CartesianViewBox;
+/**
+ * Will add all the properties required to render all individual Recharts components into a React Context.
+ *
+ * If you want to read these properties, see the collection of hooks exported from this file.
+ *
+ * @param {object} props CategoricalChartState, plus children
+ * @returns {ReactElement} React Context Provider
+ */
+export const ChartLayoutContextProvider = (props: {
+  state: CategoricalChartState;
+  children: ReactNode;
   clipPathId: string;
+}) => {
+  const {
+    state: { xAxisMap, yAxisMap, offset },
+    clipPathId,
+    children,
+  } = props;
+
+  const viewBox = calculateViewBox(offset);
+
+  /*
+   * This pretends to be a single context but actually is split into multiple smaller ones.
+   * Why?
+   * Because one React Context only allows to set one value.
+   * But we need to set multiple values.
+   * If we do that with one context, then we force re-render on components that might not even be interested
+   * in the part of the state that has changed.
+   *
+   * By splitting into smaller contexts, we allow each components to be optimized and only re-render when its dependencies change.
+   *
+   * To actually achieve the optimal re-render, it is necessary to use React.memo().
+   * See the test file for details.
+   */
+  return (
+    <XAxisContext.Provider value={xAxisMap}>
+      <YAxisContext.Provider value={yAxisMap}>
+        <ViewBoxContext.Provider value={viewBox}>
+          <ClipPathIdContext.Provider value={clipPathId}>{children}</ClipPathIdContext.Provider>
+        </ViewBoxContext.Provider>
+      </YAxisContext.Provider>
+    </XAxisContext.Provider>
+  );
+};
+
+export const useClipPathId = (): string | undefined => {
+  return useContext(ClipPathIdContext);
 };
 
-export type ChartLayoutContextType = ChartLayoutContextInitializing | ChartLayoutContextTypeLoaded;
+function getKeysForDebug(object: Record<string, unknown>) {
+  const keys = Object.keys(object);
+  if (keys.length === 0) {
+    return 'There are no available ids.';
+  }
+  return `Available ids are: ${keys}.`;
+}
+
+export const useXAxisOrThrow = (xAxisId: string | number): XAxisProps => {
+  const xAxisMap = useContext(XAxisContext);
+
+  invariant(xAxisMap != null, 'Could not find xAxisMap; are you sure this is rendered inside a Recharts context?');
```

Comment:
this is what we're telling the user. What does the user know about `xAxisMap` or Recharts context, our internal implementation? I think invariants should be directed as close to what a user can do to fix something as possible. Maybe the wording of this could be adjusted just a bit depending on when it actually happens? Same with Y below

Reply:
Updated the error message